storage: remove the concept of "dead replicas"

tbg opened the pull request:
First commit from #34987 (it just happens to touch the code here).

A long time ago, we introduced a mechanism whose intention was to allow to
detect corruption individually at a Replica level. That is, a Replica may
be found corrupt, and this would be communicated to the cluster and
measures would be taken to rebalance so that the corrupt copy would be
destroyed.

This mechanism was never relevant in practice and wouldn't have worked
properly anyway. The concept of "corruption" is pretty ill-defined. In
practice, errors from certain code paths would generically lead to
"corruption" regardless of whether they really indicated a fundamental
programming error affecting all replicas, complete disk failure, you name
it.

Knowing this, we already had code in place that would simply Fatal the
process when any replica would be marked as corrupt.

This PR goes one step further and removes the feature altogether, reducing
the amount of code we'll have to maintain during refactors going forward.
Touching both the Replica destroyStatus and the allocator, this is a good
idea because this is code we expect to be touching in the near future.

Perhaps the one "regression" worth pointing out is that we do no longer
persist the "corruption" to disk. We did so previously to ensure that the
corruption status would persist over restarts. In practice, however,
"corruption" often equaled "some error from RocksDB", and included
temporary errors (out of file descriptors, NAS device unavailable).
Persisting a breakage in the cluster in that situation is a bad idea, and
even for real "corruption" (say, a block checksum mismatch) it generally
seems desirable to be able to try again after fixing the issue manually
without additionally having to reverse-engineer the persisted corruption
state.

tbg committed:
storage: move store bootstrap to own file

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34990)
<!-- Reviewable:end -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I didn't verify that you actually caught all of the dead replica usage. I assume you grep'd the code for any other usage of that term.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34990#-:-LYrpl9t6d2DennELkMJ:b4wdgg8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=petermattis

tbg committed:
storage: remove the concept of "dead replicas"

First commit from #34987 (it just happens to touch the code here).

A long time ago, we introduced a mechanism whose intention was to allow
to detect corruption individually at a Replica level. That is, a Replica
may be found corrupt, and this would be communicated to the cluster and
measures would be taken to rebalance so that the corrupt copy would be
destroyed.

This mechanism was never relevant in practice and wouldn't have worked
properly anyway. The concept of "corruption" is pretty ill-defined. In
practice, errors from certain code paths would generically lead to
"corruption" regardless of whether they really indicated a fundamental
programming error affecting all replicas, complete disk failure, you
name it.

Knowing this, we already had code in place that would simply Fatal the
process when any replica would be marked as corrupt.

This PR goes one step further and removes the feature altogether,
reducing the amount of code we'll have to maintain during refactors
going forward. Touching both the Replica destroyStatus and the
allocator, this is a good idea because this is code we expect to
be touching in the near future.

Perhaps the one "regression" worth pointing out is that we do no longer
persist the "corruption" to disk. We did so previously to ensure that
the corruption status would persist over restarts. In practice, however,
"corruption" often equaled "some error from RocksDB", and included
temporary errors (out of file descriptors, NAS device unavailable).
Persisting a breakage in the cluster in that situation is a bad idea,
and even for real "corruption" (say, a block checksum mismatch) it
generally seems desirable to be able to try again after fixing the issue
manually without additionally having to reverse-engineer the persisted
corruption state.

Release note: None

tbg commented:
The unused lint found some more code for me to delete.

bors r=petermattis
